cpp/src/arrow/stl_iterator.h

```diff
@@ -64,12 +64,13 @@ class ArrayIterator {
   // Value access
   value_type operator*() const {
     assert(array_);
-    return array_->IsNull(index_) ? value_type{} : array_->GetView(index_);
+    return array_->IsNull(index_) ? value_type{} : ValueAccessor{}(*array_, index_);
```

Comment:
Please keep an instance of ValueAccessor as a data member here and elsewhere; otherwise it must be default constructible and empty. It wouldn't (for example) be able to reference a lookup table by reference unless that table was global.